Draft implementation of libdevice

Jokeren opened the pull request:
- [x] Unary and binary elementwise ops only?
- [x] Decorator for extern functions and builtin functions
- [x] Docs and comments
- [x] Let setup.py build extern libraries automatically

ptillet commented:
Looking good so far! To answer your questions:

1. Yes, I'm thinking we'll want to replace the math tl function with ` tl.math` module that would include all `libdevice` element-wise functions. My only concern will really have to do with exponential. Right now triton uses the approximation `2^(x * log2(e))`, which is very fast. I think we will want to update Triton-IR to have 2^x instead of exp(x) as a native instruction, and do something like:
```
def exp(x, fast=True):
  if fast:
    return 2^(x * LOG2_E)
  else:
    return __nv_exp{f,d}(x)
```

2. The design of the frontend for this feature is certainly interesting. Since the MLIR rewrite won't cover the frontend, I think we should take time to think about the best design possible for external linking. Also, because `libdevice` supports both float32 and float64, we'll likely need to automatically generate dispatcher.  Let's iterate on this. I was thinking about something along the lines of
```
# triton/language/math.py
__nv_cosh  = triton.ExternalFunction("/path/to/libdevice.bc", "__nv_cosh, tl.float64)
__nv_coshf = triton.ExternalFunction("/path/to/libdevice.bc", "__nv_coshf, tl.float32)
# cosh can be used like any element-wise function
cosh = triton.ElementwiseDispatcher({tl.float32: __nv_coshf, tl.float64: __nv_cosh})
```
In the snippet above, `ElementwiseDispatcher` would output a CompilationError when called with something else than the dtypes appearing in the implementation dictionnary. It would also check the shapes of all the input arguments. This won't work with bytecode functions that have heterogeneous argument types, though, but we could have `triton.ExternalFunction` be extended to support this case.

What do you think?

Jokeren commented:
@ptillet The design looks good to me. Thanks for your comments.

Some additional questions.

> It would also check the shapes of all the input arguments.

Why do we have to check the shape of an argument of elementwise functions?

> triton.ExternalFunction

I think what you described here is more general than purely supporting libdevice functions. Instead of specifying only the input data type, can we specify the data type of all arguments in `ExternalFunction`? 
```
__nv_cosh  = triton.ExternalFunction("/path/to/libdevice.bc", "__nv_cosh", tl.float64, tl.float64)
```
Since tl.math functions wrap up `ExternalFunction`, users still do not have to care about details unless they want to link against their own external functions.

Jokeren commented on .github/workflows/integration-tests.yml:
```diff
@@ -27,6 +27,7 @@ jobs:
       - name: Install Triton
         run: |
           alias python='python3'
+          sudo apt install -y libncurses5
```
It's due to the fact that `llvm-dis` is not installed on the test node, and installing it introducing dependency on `libtinfo`. Do you have any solutions to get around? It doesn't seem to be a problem on our A100 and V100 nodes though.

ptillet replied:
It's strange. `llvm-dis` should just use the downloaded binary (e.g., `clang+llvm-14.0.0-x86_64-linux-gnu-ubuntu-18.04/bin/llvm-dis`). If it depends on tinfo, then we could probably just grab the source code here https://github.com/llvm-mirror/llvm/blob/master/tools/llvm-dis/llvm-dis.cpp and recompile it ourself without the tinfo dependency

Jokeren commented on include/triton/codegen/selection/generator.h:
```diff
@@ -209,18 +211,26 @@ class generator: public ir::visitor, public analysis::layout_visitor {
   Builder* builder_;
   Module *mod_;
 
+  std::map<std::string, std::unique_ptr<ExternLib>> extern_libs_;
```
I'll take a look. Maybe allowing users to pass another lib path and using the default extern lib path if not specified.

Jokeren commented on python/setup.py:
```diff
@@ -54,23 +55,28 @@ def get_llvm():
     def initialize_options(self):
         build_ext.initialize_options(self)
         self.base_dir = os.path.abspath(os.path.join(os.path.dirname(__file__), os.pardir))
+        self.cuda_dir = "/usr/local/cuda"
```
Alright, that actually simplifies this work by removing the need of building `llvm-dis` (and `libncurses5`).

Jokeren replied:
Now we can point out libdevice dynamically.
Providing cuda_dir here is probably fine as we just want to allow users to specify their custom CUDA home, which could be `/scratch/user/cuda`, `/opt/cuda`, or other directories. It does not conflict with PTX path since users can still use `TRITON_PTXAS_PATH` to specify the location of `ptxas`.

ptillet replied:
But if libdevice is specified dynamically, then it sounds like we should be re-using `TRITON_PTXAS_PATH` to get the cuda path, which would make specifying `cuda_dir` here obsolete. We can also provide an additional environment variable `TRITON_LIBDEVICE_PATH`. This is probably better than specifying cuda-path at install time, since most people install the package from pip.

Jokeren replied:
I was just afraid that install could fail if CUDA is not available on the default path.

`cuda_dir` can be specified even when install the package from pip, but it is kind tricky.

ptillet replied:
Ah I see now. Triton installation doesn't actually depend on system CUDA headers anymore, so `['/usr/local/cuda/include']` should be deleted rather than replaced by something more portable.

Jokeren replied:
Got you

Jokeren commented on python/test/unit/language/test_core.py:
```diff
@@ -1300,3 +1301,49 @@ def _kernel(dst):
+    try:
+        import triton.language.libdevice as libdevice
+    except ModuleNotFoundError:
+        pytest.skip("libdevice is not installed")
```
Sure, this problem will be gone if we add a libdevice.py file.

Jokeren commented on python/setup.py:
```diff
@@ -88,17 +94,32 @@ def run(self):
-        llvm_include_dir, llvm_library_dir = get_llvm()
-        extdir = os.path.abspath(os.path.dirname(self.get_ext_fullpath(ext.path)))
+    def build_external(self, ext, llvm_dir):
+        if llvm_dir == "":
```
Likewise, this problem will be gone if we add a libdevice.py file.

Jokeren commented on python/triton/language/libdevice.py:
```diff
@@ -0,0 +1,1126 @@
+
+@extern.extern
+def clz(arg0, _builder=None):
+    return extern.elementwise("libdevice", "/usr/local/cuda/nvvm/libdevice/libdevice.10.bc", [arg0, ], {(core.dtype("int32"),): ("__nv_clz", core.dtype("int32")), (core.dtype("int64"),): ("__nv_clzll", core.dtype("int32")), }, _builder)
```
@ptillet Do you want to replace `/usr/local/cuda/nvvm/libdevice/libdevice.10.bc` with `os.environ['HOME']`?

Jokeren replied:
There's also a more flexible calling convention as illustrated in the tutorial:
```
kernel[(1,)](x_tri,` y_tri, BLOCK=shape[0], extern_libs={'libdevice': lib_path})
```

ptillet replied:
Do you think `/usr/local/cuda/nvvm/libdevice/libdevice.10.bc` could be moved into a variable at the beginning of the file, i.e. `LIBDEVICE_PATH=/usr/local/cuda/nvvm/libdevice/libdevice.10.bc` ? Let's not bother with an additional environment variable yet unless it proves useful in the future. The `extern_libs` dict is cool, especially for user-defined external libraries.

Jokeren replied:
Done, thanks for the suggestion

ptillet commented:
@jansel  I think this is dangerously close to being merged and would solve a lot of your math function issues (e.g., ceil, double-precision log, etc.). Do you think you guys could try it out and provide feedback on potential shortcomings of the API? Thanks!

Jokeren commented:
@jansel 
Please refer to `tutorials/07-libdevice-function.py` for the specific usage.
I've also assembled another example that addresses one of your concerns.
https://github.com/Jokeren/triton-samples/blob/main/float64-log.py

jansel commented:
Thanks!  I'll try this out.

jansel commented:
cc @ngimel

ngimel commented on python/triton/language/libdevice.py:
```diff
@@ -0,0 +1,1353 @@
+from . import core, extern
+
+LIBDEVICE_PATH = "/usr/local/cuda/nvvm/libdevice/libdevice.10.bc"
```
the path is hardcoded here? Will this file be regenerated if the user passes a different path?

ptillet replied:
Users can pass libdevice pass on a per-kernel basis. But your comment makes me think that we do need indeed to give users the ability to provide an environment variable name that overrides this default when creating the file. We were just talking about it yesterday with @Jokeren  :)

jansel replied:
Another idea would be to bundle it with triton at build time.

Jokeren replied:
Triton's build does not depend on any CUDA path any more. 

Our idea is allowing users to pass through the path through `os.getenv("TRITON_LIBDEVICE_PATH")`. What do you think?

ptillet replied:
One thing that would help would be to make `libdevice.py` path-independent. Then, as done currently, users run the kernel with:
```
kernel[grid](..., libs={'libdevice': path..., 'some_other_lib'=other_path})
```
with the value of `libs` defaulting to:
```
libs = {'libdevice': /path/to/ptxas/../nvvm/libdevice/libdevice.10.bc'}.
```
where /path/to/ptxas uses this function https://github.com/openai/triton/blob/2bed6fc850b93854d7f411e77c9a302954604091/lib/driver/llvm.cc#L96

ptillet replied:
@Jokeren  Let's package `libdevice.10.bc`. My understanding of the EULA https://docs.nvidia.com/cuda/eula/index.html#cuda-toolkit-supplement-license-agreement is that it is allowed.

> The following CUDA Toolkit files may be distributed with Licensee Applications developed by you, including certain variations of these files that have version number or architecture specific information embedded in the file name - as an example only, for release version 9.0 of the 64-bit Windows software, the file cudart64_90.dll is redistributable.
libdevice.10.bc

Jokeren replied:
Do you mean copy `libdevice.10.bc` to a path like `triton/language`, then the default libdevice path would be `./libdevice.10.bc`?

Jokeren replied:
Fixed in the new commit

ngimel commented on python/triton/language/libdevice.py:
```diff
@@ -0,0 +1,1353 @@
+@extern.extern
+def erf(arg0, _builder=None):
+    return extern.elementwise("libdevice", LIBDEVICE_PATH, [arg0, ], {(
+        core.dtype("fp32"),): ("__nv_erff", core.dtype("fp32")), }, _builder)
```
interesting that double version didn't get generated here https://docs.nvidia.com/cuda/libdevice-users-guide/__nv_erf.html#__nv_erf, is it not actually included in the library?

Jokeren replied:
Fixed. It was a parser error. Thanks for pointing it out.

jansel commented on python/triton/language/libdevice.py:
```diff
@@ -0,0 +1,1353 @@
+@extern.extern
+def round(arg0, _builder=None):
+    return extern.elementwise("libdevice", LIBDEVICE_PATH, [arg0, ], {(core.dtype("fp32"),): ("__nv_roundf", core.dtype("fp32")), (core.dtype(
+        "fp64"),): ("__nv_round", core.dtype("fp64")), (core.dtype("fp64"),): ("__nv_llround", core.dtype("int64")), }, _builder)
```
It looks like there are duplicate entries here:
```
(core.dtype("fp64"),): ("__nv_round", ...
(core.dtype("fp64"),): ("__nv_llround", ...
```

Jokeren replied:
Fixed

jansel commented on python/triton/language/libdevice.py:
```diff
@@ -0,0 +1,1353 @@
+@extern.extern
+def roundf(arg0, _builder=None):
+    return extern.elementwise("libdevice", LIBDEVICE_PATH, [arg0, ], {(
+        core.dtype("fp32"),): ("__nv_llroundf", core.dtype("int64")), }, _builder)
```
I'd expect this to return a fp32 and map to `__nv_roundf`

Jokeren replied:
Fixed

jansel commented:
I tried this out here https://github.com/pytorch/torchdynamo/pull/570

It works and fixes correctness issues with log/round.  Thanks again!

See minor comments above.

Jokeren commented:
Hi @ptillet @ngimel , and @jansel 

Thank you for taking time to review the code!

A lot of errors are caused by the "build_extern" parser that parses the generated ll file. Unfortunately, the libdevice.10.bc file does not seem to have a well-defined pattern for grouping based on function names, arguments, and types. 

Here's a summary of the changes I've applied:

- Added support for elementwise functions that take more than two operands. For example, `norm4d`.
- Enhanced rules to handle special cases such as `round`, `rint`, `pow`, and `erf`.
- Formated `libdevice.py` in a more human-readable form.
- Packaged `libdevice.10.bc` in triton. By default, we use `libdevice.10.bc` shipped with triton.
- Fixed problems with unsigned operands. Though LLVM does not distinguish between signed and unsigned operands, triton does.

ptillet commented:
Thanks Keren!
